snooty/flutter.py

```diff
@@ -0,0 +1,250 @@
+
+        # A custom type
+        if ty not in hints:
+            hints[ty] = _generate_hint(ty, lambda ty: inner(ty, False, level))
+
+        try:
+            return pluralize(ty.__name__)
+        except AttributeError:
+            return pluralize(_get_typename(ty))
+
+    return inner(ty, False, 0), hints
+
+
+def checked(klass: Type[T]) -> Type[T]:
+    """Marks a dataclass as being deserializable."""
+    CACHED_TYPES[klass] = None
+    return klass
+
+
+class LoadError(TypeError):
+    def __init__(self, message: str, ty: type, bad_data: object) -> None:
+        super().__init__(message)
+        self.expected_type = ty
+        self.bad_data = bad_data
+
+
+class LoadWrongType(LoadError):
+    def __init__(self, ty: type, bad_data: object) -> None:
+        description, hints = english_description_of_type(ty)
+        hint_text = '\n\n'.join(hints.values())
+        if hint_text:
+            hint_text = '\n\n' + hint_text
+
+        super().__init__(
+            f'Incorrect type. Expected {description}.{hint_text}',
+            ty,
+            bad_data)
+
+
+class LoadWrongArity(LoadWrongType):
+    pass
+
+
+class LoadUnknownField(LoadError):
+    def __init__(self, ty: type, bad_data: object, bad_field: str) -> None:
+        super().__init__('Unexpected field: "{}"'.format(bad_field), ty, bad_data)
+        self.bad_field = bad_field
+
+
+def check_type(ty: Type[C], data: object, ty_module: str = '') -> C:
+    # Check for a primitive type
+    if isinstance(ty, type) and issubclass(ty, (str, int, float, bool, type(None))):
+        if not isinstance(data, ty):
+            raise LoadWrongType(ty, data)
+        return cast(C, data)
+
+    # Check for an object
+    if isinstance(data, dict) and ty in CACHED_TYPES:
+        annotations = CACHED_TYPES[ty]
+        if annotations is None:
+            annotations = {field.name: field for field in dataclasses.fields(ty)}
+            CACHED_TYPES[ty] = annotations
+        result: Dict[str, object] = {}
+
+        # Assign missing fields None
+        missing: Set[str] = set()
+        for key in annotations:
+            if key not in data:
+                data[key] = None
+                missing.add(key)
+
+        # Check field types
+        for key, value in data.items():
+            if key not in annotations:
+                raise LoadUnknownField(ty, data, key)
+
+            field = annotations[key]
+            have_value = False
+            if key in missing:
+                # Use the field's default_factory if it's defined
+                try:
+                    result[key] = field.default_factory()  # type: ignore
+                    have_value = True
+                except TypeError:
+                    pass
+
+            if not have_value:
+                result[key] = check_type(field.type, value, ty.__module__)
+
+        output = ty(**result)
+        start_line = getattr(data, '_start_line', None)
```

Comment:
Is `_start_line` different from the other fields?  It wouldn't be included in `__annotations__`?